papis/commands/update.py

```diff
@@ -86,100 +369,163 @@ def run(document: papis.document.Document,
-                else:
-                    if importer and importer.ctx:
-                        matching_importers.append(importer)
-
-        imported = papis.utils.collect_importer_data(
-            matching_importers, batch=batch, only_data=True)
-        if "ref" in imported.data:
-            logger.debug(
-                "An importer set the 'ref' key. This is not allowed and will be "
-                "automatically removed. Check importers: '%s'",
-                "', '".join(importer.name for importer in matching_importers))
-
-            del imported.data["ref"]
-
-        ctx.data.update(imported.data)
-
-        run(document, data=ctx.data, git=git)
+        if to_set:
+            run_set(ctx.data, to_set)
+
+        if to_append and success:
+            success = run_append(ctx.data, to_append, KEY_TYPES)
+
+        if to_remove and success:
+            success = run_remove(ctx.data, to_remove)
+
+        if to_drop and success:
+            run_drop(ctx.data, to_drop)
+
+        if to_rename:
+            success = run_rename(ctx.data, to_rename)
+
+        if success:
+            logger.info(
+                "Updating {c.Back.WHITE}{c.Fore.BLACK}%s{c.Style.RESET_ALL}.",
+                papis.document.describe(document),
+            )
+
+            # NOTE: use 'papis addto' to add files, so this only adds data
+            # by setting 'only_data' to True always
+            matching_importers = papis.utils.get_matching_importer_by_name(
+                from_importer, only_data=True
+            )
+
+            if not from_importer and auto:
+                for importer_cls in papis.importer.get_importers():
+                    try:
+                        importer = importer_cls.match_data(document)
+                        if importer:
+                            try:
+                                importer.fetch_data()
+                            except NotImplementedError:
+                                importer.fetch()
+                    except NotImplementedError:
+                        continue
+                    except Exception as exc:
+                        logger.exception("Failed to match document data.", exc_info=exc)
+                    else:
+                        if importer and importer.ctx:
+                            matching_importers.append(importer)
+
+            imported = papis.utils.collect_importer_data(
+                matching_importers, batch=batch, only_data=True
+            )
+            if "ref" in imported.data:
+                logger.debug(
+                    "An importer set the 'ref' key. This is not allowed and will be "
+                    "automatically removed. Check importers: '%s'",
+                    "', '".join(importer.name for importer in matching_importers),
+                )
+
+                del imported.data["ref"]
+
+            ctx.data.update(imported.data)
```

Comment:
Probably for future PRs, but we should really be smarter about how we merge the data here. E.g. if we do `papis update --drop year` and the importers add the year back, is that a good idea?

Reply:
Maybe worth doing a `update_doc_from_data_interactively` if nothing else?

Reply:
Good point! I didn't really touch this as I don't use the importers with update. Should I add a note in the code for now to work on this later?

Reply:
Yeah, adding a little note sounds sounds good!